mmdet/core/utils/earlystop_hook.py

```diff
@@ -0,0 +1,105 @@
+        when the rule is not specific:
+        1. If the key indicator is in ``self.greater_keys``, the rule will be
+           specified as 'greater'.
+        2. Or if the key indicator is in ``self.less_keys``, the rule will be
+           specified as 'less'.
+        3. Or if the key indicator is equal to the substring in any one item
+           in ``self.greater_keys``, the rule will be specified as 'greater'.
+        4. Or if the key indicator is equal to the substring in any one item
+           in ``self.less_keys``, the rule will be specified as 'less'.
+
+        Args:
+            rule (str | None): Comparison rule for best score.
+            key_indicator (str | None): Key indicator to determine the
+                comparison rule.
+        """
+        if rule not in self.rule_map and rule is not None:
+            raise KeyError(f'rule must be greater, less or None, '
+                           f'but got {rule}.')
+
+        if rule is None:
+            if key_indicator != 'auto':
+                if key_indicator in self.greater_keys:
+                    rule = 'greater'
+                elif key_indicator in self.less_keys:
+                    rule = 'less'
+                elif any(key in key_indicator for key in self.greater_keys):
+                    rule = 'greater'
+                elif any(key in key_indicator for key in self.less_keys):
+                    rule = 'less'
+                else:
+                    raise ValueError(f'Cannot infer the rule for key '
+                                     f'{key_indicator}, thus a specific rule '
+                                     f'must be specified.')
+        self.rule = rule
+        self.key_indicator = key_indicator
+        if self.rule is not None:
+            self.compare_func = self.rule_map[self.rule]
+
+    def before_run(self, runner):
+        self.by_epoch = False if runner.max_epochs is None else True
+
+    def after_train_iter(self, runner):
+        """Called after every training iter to evaluate the results."""
+        if not self.by_epoch:
+            self._do_check_stopping(runner)
+
+    def after_train_epoch(self, runner):
+        """Called after every training epoch to evaluate the results."""
+        if self.by_epoch:
+            self._do_check_stopping(runner)
+
+    def _do_check_stopping(self, runner):
+        """perform evaluation and save ckpt."""
+        if not self._should_check_stopping(runner):
+            return
+
+        if runner.rank == 0:
+            key_score = runner.log_buffer.output[self.key_indicator]
+            if self.compare_func(key_score - self.min_delta, self.best_score):
+                self.best_score = key_score
+                self.wait_count = 0
+            else:
+                self.wait_count += 1
+
+                if self.wait_count >= self.patience:
+                    self.stopped = runner.iter
+                    raise KeyboardInterrupt(
```

Comment:
You may take a look at how `CancelTrainingHook` is implemented [here](https://github.com/openvinotoolkit/mmdetection/blob/dp/nous_sdk_api/mmdet/apis/ote/extension/utils/hooks.py#L27-L54). I believe there is a way to implement this w/o explicit utilization of file system, but just passing some messages to the running processes.

Reply:
I removed the KeyboardInterrupt and instead added CancelTrainingHook.